agents/aggregator/aggregator_agent.py

```diff
@@ -18,463 +13,511 @@ class Provider:
+                self.log.info("Provider {} went stale".format(prov.address))
+                stale_provs.append(prov)
 
-        self.log.info("Creating file: {}".format(filename))
-        self.writer = core.G3Writer(filename)
+        for prov in stale_provs:
+            self.remove_provider(prov)
 
-        session_frame = self.hksess.session_frame()
-        self.writer(session_frame)
+    def write_providers_to_queue(self, clear=True, write_all=False):
+        """
+        Loop through all providers, and write their data to the frame_queue
+        if they have surpassed their frame_time, or if write_all is True.
 
-        self.write_status()
+        Args:
+            clear (bool):
+                If True, provider data is cleared after write
+            write_all (bool):
+                If true all providers are written to disk regardless of whether
+                frame_time has passed.
+        """
+        for pid, prov in self.providers.items():
+            if write_all or prov.new_frame_time():
+                self.frame_queue.put(prov.to_frame(self.hksess, clear=clear))
 
-    def write_status(self):
+    def write_queue_to_disk(self):
         """
-        Writes hksess status frame to disk and sets write_status flag to False
+        Loops through frame_queue and writes them to disk.
         """
-        if self.writer is None:
-            return
+        while not self.frame_queue.empty():
+            self.rotator.Process(self.frame_queue.get())
 
-        status = self.hksess.status_frame()
-        self.writer(status)
-        self.should_write_status = False
+    def generate_id(self):
+        """
+        Generates a unique session id based on the start_time, process_id,
+        and hksess description.
+        """
+        # Maybe this should go directly into HKSessionHelper
+        elements = [
+            (int(self.hksess.start_time), 32),
+            (os.getpid(), 14),
+            (binascii.crc32(bytes(self.hksess.description, 'utf8')), 14)
+        ]
+        agg_session_id = 0
+        for i, b in elements:
+            agg_session_id = (agg_session_id << b) | (i % (1 << b))
+        return agg_session_id
 
-    def end_file(self):
+    def run(self):
         """
-        Writes all non-empty providers to disk, clears them,
-        and then writes an EndProcessing frame
+        Runs aggregation until `stop` is called. This should run in a worker
+        thread.
         """
-        self.log.info("Ending file")
+        self.log.info("Aggregator running...")
+        self.hksess.start_time = time.time()
+        self.hksess.session_id = self.generate_id()
 
-        for pid, prov in self.providers.items():
-            if prov.frame_start_time is None:
-                continue
+        self.rotator.Process(self.hksess.session_frame())
 
-            with prov.lock:
-                frame = prov.to_frame(self.hksess)
-                prov.clear()
+        self.aggregate = True
```

Comment:
This should probably go at the very top of the function -- or at least before the Process() -- to reduce the race-condition against stop().